File: prototype-002/src/nodejs/src/components/AwaneComponent.ts
import { randomUUID } from 'crypto';
import { IAwaneComponent } from './IAwaneComponent';

export abstract class AwaneComponent implements IAwaneComponent {
    private readonly _awaneId: string;
    private readonly _awaneInterfaces: string[];
    private readonly _awaneTags: Record<string, string>;
    
    public readonly awaneLocation: string;
    public readonly awaneName: string;
    public readonly awaneVersion: string;

    constructor() {
        this._awaneId = randomUUID();
        this.awaneLocation = 'process';
        this.awaneName = '';
        this.awaneVersion = '1.0.0';
        this._awaneTags = {};
        this._awaneInterfaces = [];
    }

    get awaneId(): string {
        return this._awaneId;
    }

    get awaneInterfaces(): string[] {
        return [...this._awaneInterfaces];
    }

    get awaneTags(): Record<string, string> {
        return { ...this._awaneTags };
    }

    public addInterface(interfaceName: string): void {
        if (!this._awaneInterfaces.includes(interfaceName)) {
            this._awaneInterfaces.push(interfaceName);
        }
    }

    public awaneAs<T>(): T {
        return this as unknown as T;
    }
}

File: prototype-002/src/csharp/Awane.Core/Components/AwaneComponent.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;

namespace Awane.Core.Components
{
    public abstract class AwaneComponent : IAwaneComponent
    {
        private readonly string _awaneId;
        private readonly string[] _awaneInterfaces;
        private readonly Dictionary<string, string> _awaneTags;

        public string AwaneId => _awaneId;
        
        public string AwaneLocation { get; protected set; }
        
        public string[] AwaneInterfaces => _awaneInterfaces;
        
        public string AwaneName { get; protected set; }
        
        public string AwaneVersion { get; protected set; }
        
        public Dictionary<string, string> AwaneTags => _awaneTags;

        protected AwaneComponent()
        {
            _awaneId = Guid.NewGuid().ToString();
            AwaneLocation = "process";
            AwaneName = string.Empty;
            AwaneVersion = "1.0.0";
            _awaneTags = new Dictionary<string, string>();
            
            // Get all interfaces implemented by this class
            var interfaces = GetType()
                .GetInterfaces()
                .Where(i => !IsSystemInterface(i))
                .Select(i => i.FullName)
                .Where(name => name != null)
                .Cast<string>()
                .ToArray();
            
            _awaneInterfaces = interfaces;
        }

        public T? AwaneAs<T>() where T : class
        {
            return this as T;
        }

        private bool IsSystemInterface(Type interfaceType)
        {
            return interfaceType.Namespace != null && 
                   interfaceType.Namespace.StartsWith("System");
        }
    }
}